File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    // Recursive solution
    public IList<int> InorderTraversal(TreeNode root) {
        List<int> result = new List<int>();
        InorderTraversalHelper(root, result);
        return result;
    }
    
    private void InorderTraversalHelper(TreeNode node, List<int> result) {
        if (node == null) return;
        
        InorderTraversalHelper(node.left, result);
        result.Add(node.val);
        InorderTraversalHelper(node.right, result);
    }
    
    // Iterative solution
    public IList<int> InorderTraversalIterative(TreeNode root) {
        List<int> result = new List<int>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode current = root;
        
        while (current != null || stack.Count > 0) {
            while (current != null) {
                stack.Push(current);
                current = current.left;
            }
            
            current = stack.Pop();
            result.Add(current.val);
            current = current.right;
        }
        
        return result;
    }
}

File: 94. Binary Tree Inorder Traversal/94. Binary Tree Inorder Traversal.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;

impl Solution {
    // Recursive solution
    pub fn inorder_traversal(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = Vec::new();
        Self::inorder_traversal_helper(&root, &mut result);
        result
    }
    
    fn inorder_traversal_helper(node: &Option<Rc<RefCell<TreeNode>>>, result: &mut Vec<i32>) {
        if let Some(n) = node {
            let n = n.borrow();
            Self::inorder_traversal_helper(&n.left, result);
            result.push(n.val);
            Self::inorder_traversal_helper(&n.right, result);
        }
    }
    
    // Iterative solution
    pub fn inorder_traversal_iterative(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<i32> {
        let mut result = Vec::new();
        let mut stack = Vec::new();
        let mut current = root.clone();
        
        while current.is_some() || !stack.is_empty() {
            while let Some(node) = current {
                stack.push(node.clone());
                current = node.borrow().left.clone();
            }
            
            if let Some(node) = stack.pop() {
                result.push(node.borrow().val);
                current = node.borrow().right.clone();
            }
        }
        
        result
    }
}